Subject: [PATCH] Fix leak of file descriptors and memory.

 1. fds must be explicitly closed if it is not retained by Flattenable::unflatten().

diff --git a/include/binder/Parcel.h b/include/binder/Parcel.h
index 74e75d7..2b283be 100644
--- a/include/binder/Parcel.h
+++ b/include/binder/Parcel.h
@@ -341,7 +341,13 @@
         }
         virtual status_t unflatten(void const* buffer, size_t size, int const* fds, size_t count) {
 #ifdef STE_HARDWARE
-            return const_cast<Flattenable&>(val).unflatten(buffer, size, fds, count);
+            status_t err = const_cast<Flattenable&>(val).unflatten(buffer, size, fds, count);
+            if (err != NO_ERROR) {
+                for (size_t i = 0; i < count; i++) {
+                    close(fds[i]);
+                }
+            }
+            return err;
 #else
             return const_cast<Flattenable<T>&>(val).unflatten(buffer, size, fds, count);
 #endif
